mc-oblivious-ram/src/evictor/mod.rs

```diff
@@ -22,28 +23,113 @@ use aligned_cmov::typenum::PartialDiv;
+
 use crate::path_oram::{BranchCheckout, MetaSize};
+use rand_core::RngCore;
 
-/// Evictor trait conceptually is a mechanism for moving stash elements into
-/// the oram.
-pub trait Evictor<ValueSize, Z>
+/// Selects branches in reverse lexicographic order. Num_bits_needed corresponds
+/// to the number of possible branches that need to be explored. The iteration i
+/// corresponds to the ith branch in reverse lexicographic order.
+fn deterministic_get_next_branch_to_evict(num_bits_needed: u32, iteration: u64) -> u64 {
+    // Return 1 if the number of bits needed is 0. This is to shortcut the
+    // calculation furtherdown that would overflow, and does not leak
+    // information because the number of bits is structural information rather
+    // than query specific.
+    if num_bits_needed == 0 {
+        return 1;
+    }
+    // This is the first index at which leafs exist, the most significant digit
+    // of all leafs is 1.
+    let leaf_significant_index: u64 = 1 << (num_bits_needed);
+    let test_position: u64 =
+        ((iteration).reverse_bits() >> (64 - num_bits_needed)) % leaf_significant_index;
+    leaf_significant_index + test_position
+}
+
+/// An evictor that implements a random branch selection and the path oram
+/// eviction strategy
+pub struct PathOramRandomEvict<RngType>
+where
+    RngType: RngCore + CryptoRng + Send + Sync + 'static,
+{
+    rng: RngType,
+    number_of_additional_branches_to_evict: usize,
+    branches_evicted: u64,
+    tree_height: u32,
+}
+
+impl<RngType> BranchSelector for PathOramRandomEvict<RngType>
+where
+    RngType: RngCore + CryptoRng + Send + Sync + 'static,
+{
+    fn get_next_branch_to_evict(&mut self) -> u64 {
+        self.branches_evicted += 1;
+        1u64.random_child_at_height(self.tree_height, &mut self.rng)
+    }
+
+    fn get_number_of_additional_branches_to_evict(&self) -> usize {
+        self.number_of_additional_branches_to_evict
+    }
+}
+impl<ValueSize, Z, RngType> EvictionStrategy<ValueSize, Z> for PathOramRandomEvict<RngType>
 where
     ValueSize: ArrayLength<u8> + PartialDiv<U8> + PartialDiv<U64>,
     Z: Unsigned + Mul<ValueSize> + Mul<MetaSize>,
     Prod<Z, ValueSize>: ArrayLength<u8> + PartialDiv<U8>,
     Prod<Z, MetaSize>: ArrayLength<u8> + PartialDiv<U8>,
+    RngType: RngCore + CryptoRng + Send + Sync + 'static,
 {
-    /// Method that takes a branch and a stash and moves elements from the
-    /// stash into the branch.
     fn evict_from_stash_to_branch(
         &self,
         stash_data: &mut [A64Bytes<ValueSize>],
         stash_meta: &mut [A8Bytes<MetaSize>],
         branch: &mut BranchCheckout<ValueSize, Z>,
-    );
+    ) {
+        path_oram_eviction_strategy::<ValueSize, Z>(stash_data, stash_meta, branch);
+    }
 }
-pub struct PathOramEvict {}
-impl<ValueSize, Z> Evictor<ValueSize, Z> for PathOramEvict
+
+/// An evictor that implements a deterministic branch selection in reverse
+/// lexicographic order and the path
+/// oram eviction strategy
```

Comment:
:wrench: This wording seems off, not sure if it was meant to be something like
```suggestion
/// lexicographic order using the path oram eviction strategy
```

Reply:
Agreed, sorry.